src/utilities/params.ts

```diff
@@ -30,6 +47,10 @@ const booleanParam: ParamGetSet<unknown> = {
 
 const numberParam: ParamGetSet<unknown> = {
   get: (value, { invalid }) => {
+    if (value === '') {
+      throw invalid()
+    }
```

Comment:
Didn't expect to see this here. An empty string should only match an optional param right?

Reply:
the process for validating params is to call the getter. So if the param finder gets an empty string for a value, and the route says a param is NOT optional and uses `Number`, it's going to call this function with an empty string, get an exception thrown, and know that param validation failed

Reply:
hmm that makes sense. But I also think the user shouldn't be responsible for checking if the value is an empty string. If a param is required the regex to match it should be `+` (at least one) right? Vs an optional param the regex would be `*` (0 or more). That way if a required param is missing from the url it doesn't even match. By the time we get to user defined `Param`s I would expect there to be a non empty string to pass into the get.

Reply:
If the regex does need to match `*` for any param maybe we should create a `required` wrapper utility that is the opposite of the `optional` utility` and wrap all `Param`s in either `optional` or `required`. That way `Param` only ever deals with actual values.

Reply:
actually you're correct that the regex test does use `+`, so technically an empty string should never makes it's way to a required param. I added this to test logic from paramValidation, but the control actually happens before paramValidation so I'll update this and move the test to before paramValidation